### ISSUE TITLE
[Proposal] Support define/override context assigns for slots

### ISSUE THREAD

**Issue by msaraiva** [msaraiva/surface#118]
There are cases when we want to allow a value of a context assign to be set dynamically by the component defining the slot. For instance, the `InputsFor` component, which will be a wrapper around `inputs_for`, could work like this without any changes in the API:

```elixir
defmodule InputsFor do
  use Surface.Component

  import Phoenix.HTML.Form
  alias Surface.Components.Form

  property form, :form

  property field, :atom

  slot default, props: [:form]

  context get form, from: Form, as: :context_form

  def render(assigns) do
    form = get_form(assigns)

    ~H"""
    <div :for={{ f <- inputs_for(form, @field) }}>
      <slot :props={{ form: f }}/>
    </div>
    """
  end
end
```

Usage:

```jsx
<Form for={{ :data }}>
  <InputsFor field={{ :children }} :let={{ form: f }}>
    <TextInput form={{ f }} field="name" />
    <TextInput form={{ f }} field="email" />
  </InputsFor>
</Form>
```

It would be nice to have a way to override the value of the context's `form` assign with the value dynamically generated by `InputsFor` so we could achieve the same behaviour using contexts, writing only:

```jsx
<Form for={{ :data }}>
  <InputsFor field={{ :children }}>
    <TextInput field="name" />
    <TextInput field="email" />
  </InputsFor>
</Form>
```

## Proposal

Add a new `context` option for `slot` where we can declare which assigns will be overridden along with its configuration.

### Example

The `InputsFor` would look something like:

```elixir
defmodule InputsFor do
  use Surface.Component

  import Phoenix.HTML.Form
  alias Surface.Components.Form

  property form, :form

  property field, :atom

  # New option `context` in action
  slot default, props: [:form], context: [form: [to: Form]]

  context get form, from: Form, as: :context_form

  def render(assigns) do
    form = get_form(assigns)

    ~H"""
    <div :for={{ f <- inputs_for(form, @field) }}>
      <slot :props={{ form: f }} :context={{ form: f } }}/>
    </div>
    """
  end
end
```

The `context: [form: [to: Form]]` looks kind of ugly but I can't think of any other better way to do that. Suggestions are welcome.

**Comment by msaraiva** [comment]
Since `InputsFor` (listed in https://github.com/msaraiva/surface/issues/32) depends on this feature, I added it to milestone [v0.1.0](https://github.com/msaraiva/surface/milestone/1).

**Comment by lnr0626** [comment]
what would you think about something like:
```
  slot default, props: [form: [override_context: {Form, :form}]]
```

where each item in `props` is either an atom, or a key: value tuple that allows you to specify options for the slot prop? It would currently only accept a single option of `override_context` that has a value of the context variable that should be overridden.

it's a bit verbose. We could also use `context_set` instead of `override_context`.

**Comment by msaraiva** [comment]
@lnr0626 thanks for the suggestion. 

One thing that concerns me about using `:props` for both cases is that we would have less flexibility. For instance, keeping them separated would allow us to add/override `context` assigns without necessarily creating `:props` vars. It also makes easier to add new options without forcing us to name them using some kind of prefix, e.g. `context_` and `prop_` to indicate that the option applies to the property or to the context. The idea behind having a separate keywod list like `[to: Form]` is that it could/should accept the same options as `context set`. Currently, it's just `:to` but it might be more as we improve contexts usage.

**Comment by lnr0626** [comment]
ohhh, I completely missed the `:context={{ form: f } }}` in your example, I thought it was automatically overriding context based on the prop of the same name. after seeing that, I also see where the `to:` comes from

yes, I agree that having them separate makes sense

**Comment by msaraiva** [comment]
I have a branch with a partially working version of `InputsFor` using the suggested approach. I'll rebase it and try to make it work with the new compiler.

**Comment by msaraiva** [comment]
After working for a while on this issue yesterday, I got the feeling that the `:context` directive solution might not be the best way to implement this. One thing that concerns me is that contexts are becoming too complex. Going in this direction would give two different ways to define/initialize context assigns, one using `context set` + `init_context` and another one using the `:context` directive. It's starting to get confusing and the implementation is reflecting that. I think maybe we should try to take a step back and rethink the whole context thing.

As an initial thought, what if we had a single generic way to set/get context assigns using a built-in component or macro. Something like this (using a new component `<Context>`):

```elixir
defmodule Form do
  use Surface.Component

  @doc "Atom or changeset to inform the form data"
  property for, :any, required: true

  @doc "URL to where the form is submitted"
  property action, :string, default: "#"

  slot default, props: [:form]

  ...

  def render(assigns) do
    ~H"""
    {{ f = form_for(@for, @action, ...) }}
      <Context set={{ form: f }}>
        <slot/>
      </Context>
    <#Raw></form></#Raw>
    """
  end
end

defmodule Field do
  use Surface.Component

  @doc "The field name"
  property name, :string, required: true

  slot default, required: true

  ...

  def render(assigns) do
    ~H"""
    <div class=...>
      <Context set={{ field: @name }}>
        <slot/>
      </Context>
    </div>
    """
  end
end

defmodule TextInput do
  use Surface.Component

  @doc "An identifier for the form"
  property form, :form

  @doc "An identifier for the input"
  property field, :string

  ...

  def render(assigns) do
    ...

    ~H"""
    <Context get={{ Form, form: context_form }} get={{ Field, field: context_field }}>
      {{ text_input(@form || context_form, @field || context_field, ...) }}
    </Context>
    """
  end
end

defmodule InputsFor do
  use Surface.Component

  ...

  @doc "An identifier for the form"
  property form, :form

  @doc "An identifier for the input"
  property field, :string

  slot default, props: [:form]

  def render(assigns) do
    form = get_form(assigns)

    ~H"""
    <div :for={{ f <- inputs_for(form, @field) }}>
      <Context set={{ Form, form: f }}>
        <slot/>
      </Context>
    </div>
    """
  end
end
```

Some benefits of the proposed solution:

1. A single way to define contexts.
1. No need to declare `context set` nor `context get`.
1. No need to implement `init_context`.
1. No clash between context assigns and props. In fact, we don't have "context assigns" anymore. Anything from the context will be available as named variables, just like slot props. They will even be passed using the same mechanism. The difference is that all context variables will be automatically passed wrapped as something like `__context_vars__`, which should only be unwrapped using `<Context get={{...}}>`.
1. Variables are scoped, i.e. context vars will be only available inside a `<Context>...</Context>`
1. Simpler implementation. To be confirmed, for sure, but since we don't have to set the values as assigns anymore, I believe some of the complexity we have today will just disappear.

@lnr0626 WDYT?

**Comment by lnr0626** [comment]
At first glance I like this proposal - i definitely see how that could simplify the context handling, and like that it would support both the current use case for contexts as well as this new requirement.

WIth the `get={{...}}` and `set={{...}}` syntax, is the intention for the first element to be the module that variable is associated with and the second a key: value pair of either context var name -> binding name (similar to `:let`), or context var name -> value (similar to `:props`)? Would the module still make sense to be associated with the context var name? I'd expect so, but something to consider.

I've taken a look through where I've used context, and the one use case I have that I don't think fits super cleanly into this computed assigns. A few components I have take advantage of the ability to add assigns through context to compute some values from the properties. Using the named assign in the template is a bit clearer to me in some cases than calling a method with a few parameters, and I didn't want to create vars in the render function based on the recommendations in live view's docs. The `init_context` function also allows returning `{:error, message}` which is convenient. That being said, I think I could still accomplish the same thing with this new proposal, it would just bit a bit more syntactically involved.

I think this would simplify allowing context values to be overridden, but I'm unsure if it'll drastically simplify what we currently have. I do think it would isolate that complexity to within this `Context` component, which would still be a benefit, and it does also make the context scoping a bit clearer/more explicit.

---

A thought just occurred to me - what's the purpose for name spacing context vars with the module - i.e. `{Form, :form}` instead of just using the var name `:form` to identify it? I see the benefit of being explicit, but if we're talking about adding a capability to override context vars, another option could be to remove that module from the context var identifier. An equivalent solution would require an additional component, but should look something like this (with better naming needed):

```elixir
defmodule InputsForContextUpdater do
  use Surface.Component
  
  property form, :form

  slot default

  context set form, :form, scope: :only_children

  def init_context(assigns) do
    {:ok, form: assigns.form}
  end

  def render(assigns) do
  ~H"""
  <slot />
  """
  end
end

defmodule InputsFor do
  use Surface.Component

  import Phoenix.HTML.Form

  property form, :form

  property field, :atom

  # New option `context` in action
  slot default, props: [:form], context: [form: [to: Form]]

  context get form, as: :context_form

  def render(assigns) do
    form = get_form(assigns)

    ~H"""
    <div :for={{ f <- inputs_for(form, @field) }}>
      <InputsForContextUpdater form={{ f }}>
        <slot :props={{ form: f }} />
      </InputsForContextUpdater>
    </div>
    """
  end
end
```

The difference here is `context get` no longer accepts `from: <module>`, and `context set` would no longer associate the variable with the module. This maintains a single way to define contexts, however wouldn't change the `context get/set` or `init_context` (though I don't dislike those as i like being able to scan the top of the module to see all of those things). We'd still have the assign name clash, and the scoping would be less configurable, but the variables are all already passed along via `@__surface__.context`. I think this might also simplify some of the current implementation, though wouldn't isolate that code.

If there's still a strong reason to keep that context var -> module association this solution wouldn't be ideal.

Thoughts?

**Comment by lnr0626** [comment]
Another similar option that would allow us to keep the context var -> module association is allowing users to set the module name in `context set` within the module itself. i.e.:

```elixir
defmodule InputsForContextUpdater do
  use Surface.Component

  alias Surface.Components.Form
  
  property form, :form

  slot default

  context set form, {Form, :form}, scope: :only_children

  def init_context(assigns) do
    {:ok, form: assigns.form}
  end

  def render(assigns) do
  ~H"""
  <slot />
  """
  end
end
```

**Comment by msaraiva** [comment]
@lnr0626 I agree. I think we could just stop grouping the values per module. That would simplify things not only regarding implementation but the syntax as well. The initial idea was to avoid conflicts between components, however, I think the best approach is to give the user ways to rename the generated variable/assign instead of scoping grouping it per module. We could still handle conflicts by printing a warning when we find one. If the component's intention is to override the value, like in `InputsFor`, we can then add an option `override: true` to make it explicit and suppress the warning.

There are other changes we need to discuss due to the way change tracking works. Currently, the generated code for `slots` disables it. I'll give you more details soon. The good news is that there's a way to make it work nicely but it requires a change in the API, for instance, instead of using the passed props as variables, we'll need to use them as assigns.

**Comment by msaraiva** [comment]
I've been playing around with the `<Context>` component and I think the best approach is to represent the context as a map where the key could be either an atom (default) or a tuple (for scoped values) as suggested by @lnr0626. 

Regarding the syntax, I believe we should keep a way to add options to individual assigns. If we restrict the input of `set` to a `:map` or `:keywork` we'll lose that ability.

One way to achieve that would be to allow multiple `set` and `get` attributes. For instance:

Setting values:

```jsx
  <Context 
    set={{ :foo, "some value" }}
    set={{ :form, form, into: __MODULE__ }}
    >
    ...
  </Context>
```

Retrieving values:

```jsx
  <Context
    get={{ :foo }}
    get={{ :form, from: Form }}
    get={{ :field, from: Field, as: my_field }}
    get={{ :current_user, as: :user }}
>
    ...
  </Context>
```

Keeping a way to customize each assign will allow us to extend the API in the future. For instance, we could have something like:

```jsx
  <Context set={{ :inputs, accumulate: true }}
    >
    ...
  </Context>
```

So instead of overriding the value when using `set` we could accumulate all values (similar to module attributes with `accumulate: true`).

I think we currently don't allow multiple attributes with the same name but I guess it would be not that hard to add this feature. Maybe we'll need to add an option `property` like `property set, ..., multiple: true`

**Comment by lnr0626** [comment]
I'm on board with this proposal and I like the idea of repeating `set`/`get` - i think it'll make it easier to see what's being used within a component, and agree with your point about allowing us to extend the api down the road.

As an alternative syntax for the `:form, from: Form`, we could also just expose the fact that it's a tuple to the user by having them specify `{Form, :form}` for `get`/`set`. your examples would then look like:

```
  <Context 
    set={{ :foo, "some value" }}
    set={{ {__MODULE__, :form}, form }}
    >
    ...
  </Context>
```
and
```
  <Context
    get={{ :foo }}
    get={{ {Form, :form} }}
    get={{ {Field, :field}, as: my_field }}
    get={{ :current_user, as: :user }}
>
    ...
  </Context>
```

After typing those out, I'm not really a fan though - your suggestion of `:atom, from:  module` makes it a bit clearer which assign would be created, and feels a bit cleaner.

> I think we currently don't allow multiple attributes with the same name but I guess it would not be that hard to add this feature. Maybe we'll need to add an option property like `property set, ..., multiple: true`

That seems reasonable, and wouldn't be a large lift at all. I would expect that a property with `multiple: true` would be exposed as a `list(:prop_type)` - i.e. `property value, :map, multiple: true` would show up in the component as a `list(:map)`.

One thing to note that's more of an implementation detail, it would be convenient to be able to `set` and `get` a value on the same `<Context>`. i.e.:

```
<Context 
  set={{ :value, some_complicated_calculation(...) }} 
  get={{ :value }}
>
   <!-- something using @value a few times -->
</Context>
```

**Comment by msaraiva** [comment]
> After typing those out, I'm not really a fan though - your suggestion of `:atom, from: module` makes it a bit clearer which assign would be created, and feels a bit cleaner.

Yeah, maybe it would be even more intuitive if instead of `into/from`, we had something like `scope` for both, `set` and `get`, for instance:

```jsx
<Context set={{ :value, scope: Mod }}>
...
</Context>

<Context get={{ :value, scope: Mod }}>
...
</Context>
```

Not sure if `scope` is the best word though :)

> I would expect that a property with `multiple: true` would be exposed as a `list(:prop_type)` - i.e. `property value, :map, multiple: true` would show up in the component as a `list(:map)`.

Yup. That's what I have in mind. We could also use `accumulate: true` instead of `multiple: true` if you prefer.

> it would be convenient to be able to `set` and `get` a value on the same

Totally agree! This is much more intuitive and powerful than the old API using:

```elixir
context set foo, :string, scope: :self_and_children
```
I never felt comfortable with that thing. I think the new API is much more consistent and allow us to do much more in a better, more natural, intuitive way.

**Comment by lnr0626** [comment]
> Yeah, maybe it would be even more intuitive if instead of into/from, we had something like scope for both, set and get, for instance
> ...
> Not sure if scope is the best word though :)

I like the idea of having the same option name for `set` and `get`, I'll see if I can come up with some other suggestions beyond `scope`

> We could also use `accumulate: true` instead of `multiple: true` if you prefer

Let's do that - that would be consistent with module attributes, so shouldn't be surprising to users

**Comment by msaraiva** [comment]
Resolved in #129
